Fix forgotten transaction encoding

v0d1ch opened the pull request:
## Why

I forgot to update the encoding of transactions in some of the fields in the `ServerOutput` previously and this PR amends that.

## What

Use `lens-aeson` package to drill into and update tx encoding in the json output for all of the fields containing transaction.
We decided to use `lens-aeson` in the end since we will need to do more changes in the `ServerOutput` than we expected and the package is convenient for this job.

---

<!-- Tick off or strike-through / remove if not applicable -->
* [ ] CHANGELOG updated
* [ ] Documentation updated
* [ ] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Use lens-aeson

v0d1ch committed:
Add failing test for tx in ServerOutput

v0d1ch committed:
Properly encode the rest of the fields containing transactions

v0d1ch committed:
Properly encode the rest of the fields containing transactions

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -161,15 +163,46 @@ prepareServerOutput OutputCBOR response =
+          key "confirmedSnapshot" .
+          key "snapshot" .
+          _Object .
+          at "confirmedTransactions" .~ cborTxs
```
Could: factor out the transaction serialization and inline these updates

I would move the `Tx -> Text` encoding into a single where clause function and then inline `replacedSnapshotTxInPostTxOnChainFailedResponse` et al to not need to invent such long name.

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -161,15 +163,46 @@ prepareServerOutput OutputCBOR response =
+            ConfirmedSnapshot{Hydra.Snapshot.snapshot} ->
+              replacedSnapshotTxInPostTxOnChainFailedResponse (confirmed snapshot)
+        _other -> encodedResponse
     RolledBack -> encodedResponse
```
As the `key` traversals are optional.. can we just apply them one after another?

i.e. we take the `encode response` and just try to do all the lens updates in sequence. They would only apply if the object has indeed the given shape (e.g. keys `postChainTx.confirmedSnapshot.snapshot.confirmedTransactions` exists)

ch1bo replied:
Ah.. the explicit pattern match is what we want to have some sanity instead. Nevermind this suggestion then.

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -161,15 +163,46 @@ prepareServerOutput OutputCBOR response =
+          case confirmedSnapshot of
+            InitialSnapshot{} -> encodedResponse
+            ConfirmedSnapshot{Hydra.Snapshot.snapshot} ->
+              replacedSnapshotTxInPostTxOnChainFailedResponse (confirmed snapshot)
```
Coudld: inline to avoid long names

```suggestion
              encodedResponse &
          key "postChainTx" .
          key "confirmedSnapshot" .
          key "snapshot" .
          key "confirmedTransactions" .~ toJSON $ toCborHex <$> confirmed snapshot
```

assuming there is a `toCborHex :: Tx -> Text`

ch1bo reviewed (approved):
It feels very dirty to do it this way, but it's quick.

v0d1ch committed:
Refactor a bit

ch1bo committed:
Inline ServerOutput json object updates

Although this looks more redundant, each case is depending on the JSON
serialization of that particular server output. Also the "key path" of
each case is easier to review if the actual update is directly in the
corresponding code branch.

v0d1ch committed:
Remove redundant imports
